### ISSUE TITLE
Write log messages to stderr instead of stdout

### ISSUE THREAD

**Issue by letmaik** [openenclave/openenclave#3177]
(from https://github.com/lsds/sgx-lkl/pull/520)

stdout is normally buffered and if a crash happens and stdout is redirected to a file then it's very likely that any error message from OE gets swallowed, which makes error diagnosis on remote systems a pain. The work-around is to disable stdout buffering (see linked PR). Ideally though any diagnostic messages (logging) from OE should go to stderr which is unbuffered by default.

@jhand2 @achamayou

**Comment by achamayou** [comment]
@letmaik we use https://linux.die.net/man/1/stdbuf

**Comment by letmaik** [comment]
@achamayou Sure, we too, but it's bad enough I even have to think about it! :)

**Comment by achamayou** [comment]
@letmaik sorry, didn’t see the PR... I do agree though, error output ought to go to stderr!

**Comment by jhand2** [comment]
You want all logs to go to stderr or just OE_LOG_LEVEL_ERROR and OE_LOG_LEVEL_FATAL?

**Comment by letmaik** [comment]
All log messages. See also https://linux.die.net/man/3/stderr. stderr is used for diagnostic and error messages. This way messages can easily be ignored by redirection.

**Comment by achamayou** [comment]
The only caveat with sending a lot of stuff to an unbuffered stream (assuming that INFO and DEBUG are potentially frequent) is that it will be quite slow. Sending FAIL&ERROR to stdout is definitely good and consistent with platform usage, INFO and DEBUG are more arguable and there's a speed tradeoff there. It's always possible to unbuffer them selectively if they go to stdout, but not if everything is sent to stderr.

**Comment by letmaik** [comment]
I agree with @achamayou, it's likely a good compromise to only send warning and higher to stderr, and the rest to stdout.

**Comment by jhand2** [comment]
>  it's likely a good compromise to only send warning and higher to stderr, and the rest to stdout

Ok I'm on board with that

**Comment by dthaler** [comment]
I'm going to argue for a different approach: everything goes to stderr but the log level should be a configurable option.  E.g., unless you set the log level to informational, informational messages go to /dev/null.   Keep stdout clean and only put stuff there that the app actually wants there.

**Comment by letmaik** [comment]
@dthaler I think @achamayou's concern is that if you *do* need to enable verbose logging for debugging purposes then sending it to stderr would be a bottleneck since the log volume is immense. I don't have numbers but apparently it's magnitudes slower, so I could imagine that getting a log dump could increase the time from a few seconds to minutes, just because stderr is unbuffered. I'll let @achamayou back that up.

**Comment by letmaik** [comment]
And @dthaler, the log level is already configurable.

**Comment by anakrish** [comment]
If the issue is that messages are lost upon a crash, is it possible to flush the stdout just prior to abnormal program termination via handling a signal or vectored exception?

**Comment by letmaik** [comment]
If you have memory corruption then you can't expect anything useful to function. I would not complicate it further.

**Comment by anakrish** [comment]
Fair point.

**Comment by anakrish** [comment]
I am leaning towards Dave's approach of logging to stderr by default. That keeps stdout clean.
